`__cleanAttrs`: Clean packages that don't expose all internals

roberth opened the pull request:
###### Description of changes

**What**

Clean up the attribute zoo on packages, opt-in.

**Why**

Automatically exposing internals blurs the line between what's expected to be stable and what's not. Let's not do that.

**How**

Call `derivationStrict` instead of `derivation`. `derivation` is backed by `derivationStrict` anyway.

When you specify `__cleanAttrs = true;` in a `mkDerivation` call, you won't be exposing internals.

**What else**

This PR improves memory by some 1-2% and cpu time also improves.

Internals are still accessible via `overrideAttrs`, `lib.inspectMkDerivationArgs` and `pkgs.inputDerivation`.

This relates weakly to a proposed improvement to make Nix take outputs from package attrsets instead of store derivations, as both this PR and the proposal work towards a more accurate as an expression-level package concept.
 - https://github.com/NixOS/nix/issues/6507

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
make-derivation.nix: Inline the call to derivation

roberth committed:
lib.extendDerivation: Allow derivationStrict-style simple outputs

roberth committed:
make-derivation.nix: Remove intermediate derivation output wrappers

roberth committed:
mkDerivation: Add __cleanAttrs ? false

This removes unnecessary implementation details from the package
attrset.

roberth committed:
mkDerivation: Do not expose passthru input attribute when __cleanAttrs

roberth committed:
mkDerivation: Remove inputDerivation from __cleanAttrs packages

roberth committed:
mkDerivation: Add name and version to __cleanAttrs packages

roberth committed:
lib.inspectMkDerivationArgs: init, alternative to .drvAttrs for __cleanArgs packages

roberth committed:
pkgs.inputDerivation: Add pkg.inputDerivation alternative for __cleanAttrs packages

roberth committed:
mkDerivation: Add outputs to __cleanAttrs packages

roberth committed:
lib.hydraJob: Accept __cleanAttrs packages

roberth committed:
hello: Expose src

roberth committed:
mkDerivation: Keep legacy name semantics

This avoids a mass-rebuild as many expressions currently use the
name+version pair instead of just a name.

The old meaning of name is based on the abused "name part" of store
paths, but there is no real reason for the attrset to use this odd
definition of name, except perhaps strong backward compatibility.

roberth committed:
make-derivation.nix: Remove a `//`

roberth committed:
make-derivation.nix: Remove a `let`

roberth committed:
cutomisation.nix: Remove a `//`

roberth committed:
lib/customisation.nix: Refactor: convert if to or

Should be faster.

roberth committed:
make-derivation.nix: Inline genAttrs

Hoping to squeeze some performance out of this.

roberth committed:
make-derivation.nix: Save another `//`

Hoping to squeeze some performance out of this.

Artturin commented:
> Internals are still accessible via overrideAttrs, lib.inspectMkDerivationArgs and pkgs.inputDerivation.

How about finalAttrs?

winterqt commented:
Do we want to make `cleanAttrs` opt-out any time soon, or will it take a lot of effort (do a lot of things in-tree depend on internal attrs)?

roberth commented:
> > Internals are still accessible via overrideAttrs, lib.inspectMkDerivationArgs and pkgs.inputDerivation.
> 
> In `mkDerivation(finalAttrs:` too?

 - `finalAttrs` and `prevAttrs` behave as before, except `finalAttrs.finalPackage` is smaller, as is to be expected because it is defined to match the return value of `mkDerivation`.

 - https://github.com/NixOS/nixpkgs/pull/211685 sets more precedent for extending `finalAttrs` with more special attributes is useful and/or necessary, so I can imagine adding the complete `drvAttrs` there, if not cost prohibitive for some reason. I think this would be an extra feature that does not have to be implemented here.

> Do we want to make `cleanAttrs` opt-out any time soon (with perhaps defaults for allowing `pname`, `version`, and `src` to passthru), or will it take a lot of effort (do a lot of things in-tree depend on internal attrs)?

I don't think we'll do this anytime soon, if at all. Applying that change too soon would mean that we fail to pay attention to the actual usage of some packages' attributes, which would be bad for users. `cleanAttrs` may never be suitable for being enabled by default in `mkDerivation`, but a successor to `mkDerivation` may enable it by default or its only behavior. The existence of `cleanAttrs` provides an anchor point for conversations about such a successor, and helps with experimentation.

SuperSandro2000 commented:
So we should document that it should never be set in any package in nixpkgs without strong reasoning?

SuperSandro2000 commented on pkgs/applications/misc/hello/default.nix:
```diff
@@ -18,6 +18,8 @@ stdenv.mkDerivation (finalAttrs: {
 
   doCheck = true;
 
+  __cleanAttrs = true;
```
then we should also not set it here

SuperSandro2000 commented on pkgs/applications/misc/hello/default.nix:
```diff
@@ -30,6 +32,9 @@ stdenv.mkDerivation (finalAttrs: {
   passthru.tests.run = callPackage ./test.nix { hello = finalAttrs.finalPackage; };
 
+  # We allow `hello.src` to be used in tests and examples.
+  passthru.src = finalAttrs.src;
```
and I think we should move all those tests out of hello and not abuse it as a testing package.

Artturin replied:
I think it's a good usage for hello

SuperSandro2000 replied:
hello is used in many guides to demonstrate how all of the nix tools work. It is great for this because it is small and still can do something. Also it is probably not installed on most systems. If we move all the tests out of it we know that those guides continue to work and can test whatever because it will not affect those.

roberth replied:
I've separated the fancy stuff into a new package `salve-mundi` based on `hello`. This way we can have our cake and eat it too. I've already removed some unnecessary bits from `hello`. I think it's at a nice level now, still avoiding `rec` and adding a test, but not overly complicated.

roberth committed:
hello: Simplify

roberth committed:
salve-mundi: init

A very fancy wrapped GNU hello, showcasing various Nixpkgs features.

roberth commented:
> So we should document that it should never be set in any package in nixpkgs without strong reasoning?

Making the interface of a package explicit helps with maintenance by reducing the surface area that a maintainer has to worry about. This also benefits users, as cleanAttrs starts the conversation about what the public interface should be. Furthermore, we can expect a further performance benefit from having smaller attrsets when applied at a larger scale.
For a single package, I don't see what specifics would make this reasoning stronger, making "strong reasons" a rather tough requirement.

roberth commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -577,14 +586,34 @@ lib.extendDerivation
+        inherit outputName outputs;
+        inherit (strict) drvPath;
+        inherit (drvAttrs) name;
+        ${if attrs.__cleanAttrs or false then null else "drvAttrs"} = drvAttrs;
```
```suggestion
        ${if attrs.__cleanAttrs or false then "internals" else "drvAttrs"} = drvAttrs;
```

Realistically, I think we'll want easier access than having to call a function to extract these bits. Having to go through a function means that the CLI becomes kind of useless, and same for tab completion in the repl.
"`internals`" does capture the intent of the attribute better: if you're going to use them, you're going to be affected by changes to internals, which are more likely than changes to the other attributes, which are implied to be a fairly stable interface. The name also opens up the possibility of exposing internals that aren't necessarily derivation arguments, but things that are currently in inaccessible `let` bindings.

DavHau replied:
If the content of `internals` und `drvAttrs` would be the same, maybe just keep `drvAttrs` as is? Changing the attr name introduces backward incompatibility.

DavHau commented on lib/derivations.nix:
```diff
@@ -98,4 +98,7 @@ in
       meta = args.meta or checked.meta;
     } // passthru;
+
+  inspectMkDerivationArgs = drv:
+    (drv.overrideAttrs (o: { passthru.__inspectArgs = o; })).__inspectArgs;
```
Not sure if relevant for this PR, but  sometimes it is useful to extract arguments of other layers of package functions. Therefore a more generic implementation could be useful here:
```nix
inspectDerivationArgsVia = funcName: drv:
    (drv.${funcName} (o: { passthru.__inspectArgs = o; })).__inspectArgs;
    
inspectMkDerivationArgs = inspectDerivationArgsVia "overrideAttrs";
```

roberth replied:
Not needed anymore, since we have `internals`. I don't think the extracted function would generalize particularly well, and I think we're trying to move away from layers of functions anyway. It's an interesting idea though and something like this could be helpful for `pkg.internals` as well, but we'll be free to change those values as we please anyway. That's kind of the point :)

DavHau commented:
> So we should document that it should never be set in any package in nixpkgs without strong reasoning?

If I understand correctly `__cleanAttrs` can be used for newly added packages without much concern because no usage exists that could break.

roberth committed:
Revert "pkgs.inputDerivation: Add pkg.inputDerivation alternative for __cleanAttrs packages"

Do not add this until we've decided what inputDerivation is
and how it fits in with cleanAttrs and the possible devShell
attribute
 - https://github.com/NixOS/nixpkgs/pull/206728

This reverts commit 99e1be309a70bfb8476b1892b8d102065860b2be.

roberth committed:
make-derivation.nix: Use formal for __cleanAttrs

roberth committed:
make-derivation.nix: Do expose `internals` when `__cleanAttrs`

Realistically, we'll want easier access than having to call a
function to extract these bits. Having to go through a function
means that the CLI becomes almost useless for troubleshooting,
and the same for tab completion in the repl.

roberth committed:
Revert "lib.inspectMkDerivationArgs: init, alternative to .drvAttrs for __cleanArgs packages"

This was too unergonomic to be the only method for retrieving
these. By introducing the `internals` attribute for this purpose
instead, we restore easy access without setting any wrong
expectations regarding the stability of the contents.

This reverts commit 01eea717c474d25b1636f9ee93c9f7bf95907fc6.

roberth commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -314,7 +315,9 @@ else let
     }) // lib.optionalAttrs __structuredAttrs { env = checkedEnv; } // {
+      # NOTE: keep in sync with trivial-builders.nix#inputDerivation
       builder = attrs.realBuilder or stdenv.shell;
+      # NOTE: keep in sync with trivial-builders.nix#inputDerivation
```
It was disproportionally difficult to share a function to DRY this. `make-derivation.nix` isn't set up to have access to a lot of things. I've resolved it be deleting the extra function, because there's no immediate need and `inputDerivation` has an uncertain future anyway.

roberth committed:
salve-mundi.nix: Explain what the file is for

roberth committed:
doc/stdenv: Document __cleanAttrs

roberth committed:
doc/stdenv: Document __structuredAttrs

roberth committed:
pkgs.tests.stdenv: Test __cleanAttrs

roberth committed:
doc/rl-2305: Add __cleanAttrs

fricklerhandwerk reviewed (commented):
Docs LGTM

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-03-13-nixpkgs-architecture-team-meeting-32/26287/1

roberth committed:
mkDerivation: Add `__cleanAttrs = "warn";`

The implementation with dynamic attributes is a little weird, but
it performs better by avoiding some attrset copying and sorting, `//`.

infinisil commented:
I didn't mention this in the team meeting, but I don't think we need `__cleanAttrs = "warn"`, I think we can just make `__cleanAttrs = true` warn for one release, then throw for one release, then actually remove the attributes in another release. We'd still get the same benefits (just not immediately), it adds little complexity to the code, and could probably prevent a lot of frustration depending, especially if we'd like to make `__cleanAttrs` widely used.

roberth commented:
> I think we can just make `__cleanAttrs = true` warn for one release

Then every package that wants to opt in would have to do it in the first release. I don't think that's a good approach, and the cost of `"warn"` is low enough.

I'd like to avoid blowing up the scope and merge this soon. It adds the mechanism, but avoids changing the policy, so that we can
 - start applying the solution on a small scale
 - start solving the engineering problem
 - gain data and experience to inform policy decisions

SuperSandro2000 commented:
> Making the interface of a package explicit helps with maintenance by reducing the surface area that a maintainer has to worry about. This also benefits users, as cleanAttrs starts the conversation about what the public interface should be. Furthermore, we can expect a further performance benefit from having smaller attrsets when applied at a larger scale.
> For a single package, I don't see what specifics would make this reasoning stronger, making "strong reasons" a rather tough requirement.

For me this translates into more work and discussions. Sometimes I know that the solution I am drafting is hacky and I decided for myself that it is better than the available alternatives. I don't want to contact another person to expose some variable(s) for that though. I personally would want a setting to just disable cleanAttrs everywhere and give me access to everything, as I don't really care if something is marked as an interface or not.
Or do we now need to start doing reflections?

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2305.section.md:
```diff
@@ -165,6 +165,10 @@ In addition to numerous new and upgraded packages, this release has the followin
 
+- `mkDerivation` does not have to leak implementation details anymore.
+
+  Passing [`__cleanAttrs = true`](https://nixos.org/manual/nixpkgs/unstable/#var-__cleanAttrs) to `mkDerivation` returns a minimal set of package attributes that can be extended via [`passthru`](https://nixos.org/manual/nixpkgs/unstable/#var-stdenv-passthru). This reduces confusion for package users and allows for confident use of non-standard attributes provided by a package. It also gives packa ...
```
I really don't like the wording of this. It encourages people to just set cleanAttrs for all their packages, only to later get feedback from people that they need x, y, z back.

roberth replied:
How about this?

```suggestion
  Passing [`__cleanAttrs = true`](https://nixos.org/manual/nixpkgs/unstable/#var-__cleanAttrs) to `mkDerivation` returns a minimal set of package attributes that can be extended via [`passthru`](https://nixos.org/manual/nixpkgs/unstable/#var-stdenv-passthru). For new packages, this reduces confusion for package users and allows for confident use of non-standard attributes provided by a package. It also gives package authors more confidence in modifying the builder environment without breaking consumers of their package, and it is slightly more efficient. It may be applied to existing packages with great care, as other users may unknowingly rely on internals. To uncover such usages without breaking their code, you may first pass `__cleanAttrs = "warn"` instead.
```

SuperSandro2000 replied:
That sounds better. 👍🏼

SuperSandro2000 commented on pkgs/applications/misc/hello/salve-mundi.nix:
```diff
@@ -0,0 +1,98 @@
+
+  __cleanAttrs = true;
+  passthru.exe = lib.getExe finalAttrs.finalPackage;
+  # We allow `hello.src` to be used in tests and examples, despite __cleanAttrs
+  passthru.src = finalAttrs.src;
```
IMO src should always be passed through. eg you sometimes want to (re-)build it and as I understand it, that would after that no longer be possible which would suck.

roberth replied:
> For me this translates into more work and discussions. Sometimes I know that the solution I am drafting is hacky and I decided for myself that it is better than the available alternatives. I don't want to contact another person to expose some variable(s) for that though.

You'll still be able to do this, by using attributes such as `.internals.src`. If you don't think your use case should be explicitly supported by the package, you don't need to do anything.

Also if you're very confident about your use case, you could just open a PR. It doesn't have to some sort of drawn out discussion.

> I personally would want a setting to just disable cleanAttrs everywhere and give me access to everything, as I don't really care if something is marked as an interface or not. Or do we now need to start doing reflections?

What do you mean with reflections?

I'm not in favor of such a global flag, because it is not necessary and it will make some expressions that do or do not use the flag incompatible. (Configuration should always be a last resort)

> For example this would majorly break nix-update and complicate it.
> 
> https://github.com/Mic92/nix-update/blob/f7eba589cc99a0cb6f0bab298e194c7c5705be13/nix_update/eval.py#L106-L118

It'd be fine for `nix-update` to peek inside the `internals` and then carry on as it did.

It would be nicer if the language integrations and `nix-update` agreed on some small but general interface, such as a package attribute, so that `nix-update` isn't coupled to each of the language integrations. More concretely, if the language integrations provided a default `updateScript` attribute, `nix-update` would be trivial. That's a separate project though that I don't really want to delve into here, but I think it shows that `internals` highlights the places where the design could be improved.

roberth replied:
`src` -> `internals.src` should be fine for individual "override" packages you write.

There is no general way to get all sources in using just the expression language, so we don't lose the capability to generically retrieve all sources, because we never had that in the language. Simply put, not all packages have a `src`, and not all sources appear in `src`, and you can hide "anything" in a dependency of a derivation a string context.
This task is better achieved by analyzing the `.drv` graph, you can find _all_ the derivation input sources (file references and builtin fetcher results) and fixed output derivations. You may even do this from within a derivation.

roberth commented:
> For me this translates into more work and discussions. Sometimes I know that the solution I am drafting is hacky and I decided for myself that it is better than the available alternatives. I don't want to contact another person to expose some variable(s) for that though.

You'll still be able to do this, by using attributes such as `.internals.src`. If you don't think your use case should be explicitly supported by the package, you don't need to do anything.

Also if you're very confident about your use case, you could just open a PR. It doesn't have to some sort of drawn out discussion.

> I personally would want a setting to just disable cleanAttrs everywhere and give me access to everything, as I don't really care if something is marked as an interface or not. Or do we now need to start doing reflections?

What do you mean with reflections?

I'm not in favor of such a global flag, because it is not necessary and it will make some expressions that do or do not use the flag incompatible. (Configuration should always be a last resort)

> For example this would majorly break nix-update and complicate it.
> 
> https://github.com/Mic92/nix-update/blob/f7eba589cc99a0cb6f0bab298e194c7c5705be13/nix_update/eval.py#L106-L118

It'd be fine for `nix-update` to peek inside the `internals` and then carry on as it did.

It would be nicer if the language integrations and `nix-update` agreed on some small but general interface, such as a package attribute, so that `nix-update` isn't coupled to each of the language integrations. More concretely, if the language integrations provided a default `updateScript` attribute, `nix-update` would be trivial. That's a separate project though that I don't really want to delve into here, but I think it shows that `internals` highlights the places where the design could be improved.

SuperSandro2000 commented:
> You'll still be able to do this, by using attributes such as `.internals.src`.

Totally forgot about that. 😓 Then this should be okay to megre.

infinisil commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -542,11 +548,37 @@ else let
   validity.handled
-  ({
+  (
+    lib.optionalAttrs (__cleanAttrs != true) (maybeWarnCleanAttrs {
```
This removes `inputDerivation` when `__cleanAttrs == true`, I'm not sure if that makes sense, since `.inputDerivation` is meant to be a public stable interface. While it might not be needed anymore since [`mkShell` is now buildable](https://github.com/NixOS/nixpkgs/pull/153194), I'm not sure if this should be decided here.

infinisil commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -577,14 +609,37 @@ lib.extendDerivation
+  (
+      (
+        if __cleanAttrs == true
+        then builtins.intersectAttrs { version = null; } drvAttrs
```
Why is version propagated? I guess it's somewhat important and stable enough. Should definitely be a comment here

infinisil commented on pkgs/applications/misc/hello/salve-mundi.nix:
```diff
@@ -0,0 +1,98 @@
+      touch $out
+    '';
+
+    restore-default-greeting = callPackage ./test.nix {
+      hello = finalAttrs.finalPackage.overrideAttrs (o: {
+        passthru = o.passthru // {
+          greeting = null;
+        };
+      });
+    };
```
What's this for?

blaggacao replied:
not everyone speaks latin :smile:

roberth replied:
That is a good point. It could come across as elitist, especially in this context now that I think about it more.
Something like Esperanto would be nicer, or Lojban, or Tolkien's Elvish. Or maybe it should just be `hello_with_bells_and_whistles`, although I don't know if `_` for variations should be followed by snake case or camelCase.
Why is this hard...

May the first confirmation win:
 - `suilad-ambar`: is this a correct Elvish package name?
 - `coi-munje`: is this a correct Lojban package name?

I don't like the Esperanto version because it's basically Romance language vocabulary. Nothing wrong with that, but if it's not going to be English, I don't want to bias it. Sue me. ;)

infinisil reviewed (changes requested):
I'm a bit worried that people will start using `__cleanAttrs = true` on existing packages, breaking a lot of things without warnings downstream (and for every user complaining about an attribute gone missing there's 10 more in the back that never reported it). We'd have to trust reviewers to not merge such changes. Better of course would be CI, but that adds a lot of complexity.

Other than that this looks pretty good to me, I like the salve mundi!

fricklerhandwerk commented on doc/stdenv/stdenv.chapter.md:
```diff
@@ -362,6 +364,42 @@ Unless set to `false`, some build systems with good support for parallel buildin
 
+#### `__cleanAttrs` {#var-__cleanAttrs}
+
+By default, `mkDerivation` will expose its arguments in the returned package attribute set. This is unnecessary and leads to some confusion and doubt.
```
```suggestion
By default, `mkDerivation` will expose its arguments in the returned package attribute set.
This is often unnecessary and may lead to confusion about what's relevant for consumers.
```

We should soften that a bit, because I don't see evidence for the originally definitive-sounding statement.

fricklerhandwerk commented on doc/stdenv/stdenv.chapter.md:
```diff
@@ -362,6 +364,42 @@ Unless set to `false`, some build systems with good support for parallel buildin
+
+##### What to do when warned {#warning-package-attr-impl-detail}
+
+If you encounter the warning `The attribute ... of package ... is an implementation detail`, you are invited to help us explicitly support ways in which a package attribute set may used, so that Nixpkgs contributors and users will be aware of your use case.
```
This could be phrased to be more immediately actionable.

```suggestion
If you encounter the warning `The attribute ... of package ... is an implementation detail`, please make Nixpkgs contributors and users aware of your particular use case, and help providing explicit ways to use package attribute sets, by opening an [issue](https://github.com/NixOS/nixpkgs/issues) or (https://github.com/NixOS/nixpkgs/pulls).
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/zurich-23-05-zhf-hackathon-and-workshop-report/29093/1

roberth commented:
https://github.com/NixOS/nixpkgs/pull/271241#issuecomment-1834596148 shows that access to values computed by `mkDerivations` can be important for the package definition itself. Ideally this would not go through the public interface, but I'm not sure that I would want to impose `finalAttrs.finalPackage.internals.doCheck` as a best practice...

Perhaps it'd be better to avoid cleanAttrs and instead move to a module-like situation, where the functions are flattened to a single fixed-point, and the attrsets are unflattened so that what used to be layers between the functions are just groups of attributes.

```nix
mk2Derivation({ pythonAttrs, setup, drvAttrs, public }:
  setup.configureFlags = ... (optional drvAttrs.doCheck "--enable-tests") ...;
  # no more passthru
  public.tests = callPackage ./tests { mypkg = public; };
)
```

Is the overlay pattern (with 1 layer of hardcoded merging) good enough for this, or do we need more module system-like features...

roberth commented:
> automatic merging of values would be nice, but maybe it's not a must for the start.

Not merging those "top level" attributes, of which we have only few, will make any sort of overriding incredibly needlessly labor intensive, while gaining very little in flexibility or performance (if at all).
I believe we may allow a restricted form of module system-like merging, as the merging [implemented](https://gist.github.com/roberth/940dff88ca5f5f95949dc309dbe60a65#file-minimod-nix-L126) in minimod appears quite cheap, especially if we can get rid of priorities, thanks to the implicit priority encoded by overlay layer ordering.

> `__class`

We already have `_class` in the module system (newish feature) which is named too similarly, yet must be quite different in interpretation to coexist.

Whether similarity - let alone compatibility - with the module system is a valid requirement remains to be seen, because we can not degrade performance by any significant amount, due to the scale of a typical package's dependency closure.

roberth commented:
This idea was good, but `mkDerivation` can not sustain it.
It appeared as low hanging fruit, but it was ever so slightly _too much_, as exemplified by https://github.com/NixOS/nixpkgs/pull/217243#issuecomment-1835219430.
